managed/services/uievents/uievents.go

```diff
@@ -0,0 +1,262 @@
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package uievents provides facility to store UI events.
+package uievents
+
+import (
+	"context"
+	"encoding/json"
+	"sort"
+	"time"
+
+	"github.com/HdrHistogram/hdrhistogram-go"
+	pmmv1 "github.com/percona-platform/saas/gen/telemetry/events/pmm"
+	"github.com/sirupsen/logrus"
+
+	uievents "github.com/percona/pmm/api/uieventspb"
+	"github.com/percona/pmm/managed/services/telemetry"
+)
+
+const (
+	cleanupAfterHours    = 30
+	cleanupCheckInterval = 1 * time.Minute
+)
+
+// Service provides facility for storing UI events.
+type Service struct {
+	l               *logrus.Entry
+	lastCleanup     time.Time
+	dashboardUsage  map[string]*DashboardUsageStat
+	componentsUsage map[string]*ComponentsUsageStat
+
+	uievents.UnimplementedUIEventsServer
+}
+
+type DashboardUsageStat struct {
+	title    string
+	uid      string
+	useCount int32
+	loadTime *hdrhistogram.Histogram
+}
+
+type ComponentsUsageStat struct {
+	uid      string
+	useCount int32
+	loadTime *hdrhistogram.Histogram
+}
+
+// New returns platform Service.
+func New() (*Service, error) {
+	l := logrus.WithField("component", "platform")
+
+	s := Service{
+		l:               l,
+		dashboardUsage:  make(map[string]*DashboardUsageStat),
+		componentsUsage: make(map[string]*ComponentsUsageStat),
+		lastCleanup:     time.Now(),
+	}
+
+	return &s, nil
+}
+
+// ScheduleCleanup schedules internal clean for internal UI event storage.
+func (s *Service) ScheduleCleanup(ctx context.Context) {
+	go func() {
+		ticker := time.NewTicker(cleanupCheckInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				s.cleanupIfNeeded()
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+func (s *Service) FetchMetrics(ctx context.Context, report *telemetry.Config) ([]*pmmv1.ServerMetric_Metric, error) {
+	var metrics []*pmmv1.ServerMetric_Metric
+
+	metrics = s.processDashboardStat(metrics)
+	metrics = s.processComponentsStat(metrics)
+
+	return metrics, nil
+}
+
+func (s *Service) processDashboardStat(metrics []*pmmv1.ServerMetric_Metric) []*pmmv1.ServerMetric_Metric {
+	type Stat struct {
+		TopDashboards         []string `json:"top_dashboards"`
+		SlowDashboardsP95_1s  []string `json:"slow_dashboards_p95_1s"`
+		SlowDashboardsP95_5s  []string `json:"slow_dashboards_p95_5s"`
+		SlowDashboardsP95_10s []string `json:"slow_dashboards_p95_10s"`
+	}
+
+	if len(s.dashboardUsage) > 0 {
+		dashboardStat := &Stat{
+			SlowDashboardsP95_1s:  []string{},
+			SlowDashboardsP95_5s:  []string{},
+			SlowDashboardsP95_10s: []string{},
```

Comment:
just a question, why do we set value for only 3 of 4 fields?

Reply:
sorry, didn't get the question

Reply:
this struct has 4 fields, but here we fill only 3 of them